pkg/apis/auth/v1alpha1/types.go

```diff
@@ -22,10 +25,447 @@ type ClusterAuthentication struct {
+	URL string `json:"url"`
+	// ClientID is the oauth client ID
+	ClientID string `json:"clientID"`
+	// ClientSecret is the oauth client secret
+	ClientSecret StringSource `json:"clientSecret"`
+	// LegacyOAuth2 determines that OAuth2 should be used, not OIDC
+	LegacyOAuth2 bool `json:"legacy,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// GoogleIdentityProvider provides identities for users authenticating using Google credentials
+type GoogleIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// ClientID is the oauth client ID
+	ClientID string `json:"clientID"`
+	// ClientSecret is the oauth client secret
+	ClientSecret StringSource `json:"clientSecret"`
+
+	// HostedDomain is the optional Google App domain (e.g. "mycompany.com") to restrict logins to
+	HostedDomain string `json:"hostedDomain"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// OpenIDIdentityProvider provides identities for users authenticating using OpenID credentials
+type OpenIDIdentityProvider struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CA is the optional trusted certificate authority bundle to use when making requests to the server
+	// If empty, the default system roots are used
+	CA string `json:"ca"`
+
+	// ClientID is the oauth client ID
+	ClientID string `json:"clientID"`
+	// ClientSecret is the oauth client secret
+	ClientSecret StringSource `json:"clientSecret"`
+
+	// ExtraScopes are any scopes to request in addition to the standard "openid" scope.
+	ExtraScopes []string `json:"extraScopes"`
+
+	// ExtraAuthorizeParameters are any custom parameters to add to the authorize request.
+	ExtraAuthorizeParameters map[string]string `json:"extraAuthorizeParameters"`
+
+	// URLs to use to authenticate
+	URLs OpenIDURLs `json:"urls"`
+
+	// Claims mappings
+	Claims OpenIDClaims `json:"claims"`
+}
+
+// OpenIDURLs are URLs to use when authenticating with an OpenID identity provider
+type OpenIDURLs struct {
+	// Authorize is the oauth authorization URL
+	Authorize string `json:"authorize"`
+	// Token is the oauth token granting URL
+	Token string `json:"token"`
+	// UserInfo is the optional userinfo URL.
+	// If present, a granted access_token is used to request claims
+	// If empty, a granted id_token is parsed for claims
+	UserInfo string `json:"userInfo"`
+}
+
+// UserIDClaim is used in the `ID` field for an `OpenIDClaim`
+// Per http://openid.net/specs/openid-connect-core-1_0.html#ClaimStability
+//  "The sub (subject) and iss (issuer) Claims, used together, are the only Claims that an RP can
+//   rely upon as a stable identifier for the End-User, since the sub Claim MUST be locally unique
+//   and never reassigned within the Issuer for a particular End-User, as described in Section 2.
+//   Therefore, the only guaranteed unique identifier for a given End-User is the combination of the
+//   iss Claim and the sub Claim."
+const UserIDClaim = "sub"
```

Comment:
@enj removed from `OpenIDClaims` below and added this const to keep track of it.